File: 백준/Silver/9012. 괄호/괄호.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        String[] answer = new String[t];

        for (int i = 0; i < t; i++) {
            Stack<Character> st = new Stack<>();
            answer[i] = "YES";
            for (char c : sc.next().toCharArray()) {
                if (c == '(') {
                    st.push(c);
                } else {
                    if(st.isEmpty()) answer[i] = "NO";
                    else st.pop();
                }
            }
            if(!st.isEmpty()) answer[i] = "NO";
        }
        for(String s: answer) System.out.println(s);
    }
}

File: 백준/Silver/9012. 괄호/괄호.py
def is_vps(s):
    stack = []
    for char in s:
        if char == '(':
            stack.append(char)
        elif char == ')':
            if stack:
                stack.pop()
            else:
                return "NO"
    
    if not stack:
        return "YES"
    else:
        return "NO"

n = int(input())
for _ in range(n):
    ps = input()
    print(is_vps(ps))
